eZ/Publish/Core/REST/Server/Controller/Role.php

```diff
@@ -204,30 +212,35 @@ public function updateRoleDraft($roleId, Request $request)
             )
         );
 
-        return $this->roleService->updateRoleDraft(
-            $this->roleService->loadRoleDraft($roleId),
-            $this->mapToUpdateStruct($createStruct)
-        );
+        try {
+            // First try to load the draft for given role.
+            $roleDraft = $this->roleService->loadRoleDraftByRoleId($roleId);
+        } catch (NotFoundException $e) {
+            // We might want a newly created role, so try to load it by its ID.
+            // loadRoleDraft() might throw a NotFoundException (wrong $roleId). If so, let it bubble up.
+            $roleDraft = $this->roleService->loadRoleDraft($roleId);
+        }
+
+        return $this->roleService->updateRoleDraft($roleDraft, $this->mapToUpdateStruct($createStruct));
     }
 
     /**
      * Publishes a role draft.
      *
-     * @param $roleId
+     * @param mixed $roleId Original role ID, or ID of the role draft itself
      */
-    public function publishRoleDraft($roleId, Request $request)
+    public function publishRoleDraft($roleId)
     {
-        $createStruct = $this->inputDispatcher->parse(
-            new Message(
-                array('Content-Type' => $request->headers->get('Content-Type')),
-                $request->getContent()
-            )
-        );
+        try {
+            // First try to load the draft for given role.
+            $roleDraft = $this->roleService->loadRoleDraftByRoleId($roleId);
+        } catch (NotFoundException $e) {
+            // We might want a newly created role, so try to load it by its ID.
+            // loadRoleDraft() might throw a NotFoundException (wrong $roleId). If so, let it bubble up.
+            $roleDraft = $this->roleService->loadRoleDraft($roleId);
+        }
 
-        $this->roleService->publishRoleDraft(
-            $this->roleService->loadRoleDraft($roleId),
-            $this->mapToUpdateStruct($createStruct)
-        );
+        $this->roleService->publishRoleDraft($roleDraft);
```

Comment:
The action should return something :-)

It would be something like return a `PublishedRole` value object or something, that has the Role as a property. Then the visitor would send a 200, with a Location header to the published Role (so that API users can follow the link).